src/mlpack/methods/ann/layer/parametric_relu_impl.hpp

```diff
@@ -0,0 +1,140 @@
+/**
+ * @file methods/ann/layer/parametric_relu_impl.hpp
+ * @author Prasanna Patil
+ *
+ * Definition of PReLU layer first introduced in the,
+ * Kaiming He, Xiangyu Zhang, Shaoqing, Ren Jian Sun,
+ * "Delving Deep into Rectifiers:
+ * Surpassing Human-Level Performance on ImageNet Classification", 2014
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_PRELU_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_PRELU_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "parametric_relu.hpp"
+
+namespace mlpack {
+
+template<typename MatType>
+PReLUType<MatType>::PReLUType(const double userAlpha) :
+    Layer<MatType>(),
+    userAlpha(userAlpha)
+{
+  alpha.set_size(WeightSize(), 1);
+  alpha(0) = userAlpha;
+}
+
+template<typename MatType>
+PReLUType<MatType>::PReLUType(
+    const PReLUType& other) :
+    Layer<MatType>(other),
+    userAlpha(other.userAlpha)
+{
+  alpha.set_size(WeightSize(), 1);
+  alpha(0) = userAlpha;
+}
+
+template<typename MatType>
+PReLUType<MatType>::PReLUType(
+    PReLUType&& other) :
+    Layer<MatType>(std::move(other)),
+    userAlpha(std::move(other.userAlpha))
+{
+  alpha.set_size(WeightSize(), 1);
+  alpha(0) = userAlpha;
+}
+
+template<typename MatType>
+PReLUType<MatType>&
+PReLUType<MatType>::operator=(const PReLUType& other)
+{
+  if (&other != this)
+  {
+    Layer<MatType>::operator=(other);
+    userAlpha = other.userAlpha;
+    alpha = other.alpha;
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+PReLUType<MatType>&
+PReLUType<MatType>::operator=(PReLUType&& other)
+{
+  if (&other != this)
+  {
+    Layer<MatType>::operator=(std::move(other));
+    userAlpha = std::move(other.userAlpha);
+    alpha = std::move(other.alpha);
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+void PReLUType<MatType>::SetWeights(
+    typename MatType::elem_type* weightsPtr)
+{
+  alpha = arma::mat(weightsPtr, 1, 1, false, false);
+
+  //! Set value of alpha to the one given by user.
+  // TODO: this doesn't even make any sense.  is it trainable or not?
+  // why is there userAlpha?  is that for initialization only?
```

Comment:
This looks like a slightly frustrated comment I left some time ago :smile: 

We should probably figure it out now in this PR. :)  I am not sure what the original intention of the code was---does the paper suggest that setting a controllable initial value of `alpha` is a good thing?

Reply:
Also, this is unsafe: when a user deserializes a network from disk, this will overwrite `alpha` with whatever the `userAlpha` was.  So either we need to ensure that `userAlpha` is always equal to what `alpha(0)` is when we serialize (and thus will also be equal when we deserialize), or find some other way to initialize it.

Reply:
I know I am now leaving three comments here... sorry about that.

I realized the "clean" way to do this with the abstractions we have is to override `CustomInitialize()`; take a look at `layer/layer.hpp` or `layer/batch_norm_impl.hpp` for an example.  That will allow setting the single weight of the network to `userAlpha`, and that will only be called when the network is being initialized---not when it's being just serialized or deserialized.

Reply:
According to the paper mentioned in code, Yes, `alpha` is a trainable parameter and `userAlpha` is just for initialization. I am not sure but there is no need of using `CustomInitialize()`. I used alpha just like `ratio` in `layer/dropout.hpp` which is also trainable in the last commit. Please guide me if I am not able to understand.

Reply:
[This comment](https://github.com/mlpack/mlpack/pull/3420/files#r1109265575) explains that the strategy here is incorrect.  Please read it carefully and reference relevant code (and walk through the `FFN` code) if you don't believe me.  We will need to use `CustomInitialize()`.